mirascope/core/base/toolkit.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+import inspect
+from abc import ABC
+from typing import Callable, ClassVar, ParamSpec, NamedTuple
+
+from pydantic import BaseModel, ConfigDict
+from typing_extensions import LiteralString
+
+from . import BaseTool
+from ._utils import convert_function_to_base_tool, get_template_variables
+
+_TOOLKIT_TOOL_METHOD_MARKER: LiteralString = "__toolkit_tool_method__"
+
+P = ParamSpec("P")
+
+
+def toolkit_tool(
+    method: Callable[[BaseToolKit, ...], str],
+) -> Callable[[BaseToolKit, ...], str]:
+    # Mark the method as a toolkit tool
+    setattr(method, _TOOLKIT_TOOL_METHOD_MARKER, True)
+
+    return method
+
+
+class TookKitToolMethod(NamedTuple):
+    method: Callable[[BaseToolKit, ...], str]
+    template_vars: list[str]
+    template: str
+
+
+class BaseToolKit(BaseModel, ABC):
+    """A class for defining tools for LLM call tools."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    _toolkit_tool_methods: ClassVar[list[TookKitToolMethod]]
+
+    def create_tools(self) -> list[type[BaseTool]]:
+        """The method to create the tools."""
+        tools: list[type[BaseTool]] = []
+        for method, template_vars, template in self._toolkit_tool_methods:
```

Comment:
would it be cleaner to just return a list comprehension here? For example:

```python
def create_tools(self) -> list[type[BaseTool]]:
    """The method to create the tools."""
    return [
        convert_function_to_base_tool(method, BaseTool, template.format(self=self))
        for method, template_vars, template in self._toolkit_tool_methods
    ]
```

Reply:
I think it's very good.
I love list comprehension, but I refrain from using them if they are long because they are not generally accepted.

Reply:
I agree. I think if it's too complex, best to avoid, but it's simpler to read in this case I believe